File: dlg/dialogconvert.pl
use strict;
use warnings;

my $infile = @ARGV[0];
my $outfile = 'npcdlg.txt';

open my $info, $infile or die "Could not open $infile: $!";
open my $fh, '>>', $outfile or die "Could not open file '$outfile' $!";

while( my $line = <$info>)
{
	my @values = split('\t',$line);
	my $command1 = "send o 0 speechlibtrigger who class $values[0] trigger q $values[2]";
	my $command2 = "send o 0 speechlibquote who class $values[0] quote q $values[3]\n";
	my $command3 = "send o 0 randomlibquote who class $values[0]";
	#print "0:$values[0]\n1:$values[1]\n2:$values[2]\n3:$values[3]\n----\n";
	# Check if we have a quote
	if (length $values[3])
	{
		# If we have a trigger, use command1 and command 2. If not, use command 3
		if (length $values[2])
		{
			# Add the mood if present
			if (length $values[1])
			{
				$command1 = $command1 . " mood int $values[1]";
			}
			print $fh "$command1\n$command2";
			#last if $. == 200;
		}
		else
		{
			# Add the mood if present
			if (length $values[1])
			{
				$command3 = $command3 . " mood int $values[1]";
			}
			$command3 = $command3 . " quote q $values[3]";
			print $fh "$command3\n";
		}
	}
}

close $infile;
close $outfile;

File: dlg/dialogconvert.py
# Convert from .tab file format to a sequence of admin commands.
# After recreating the world, e.g. "send o 0 recreate all" (or at initial startup),
# add NPC dialog to the speech library with these commands:
#
# python dialogconvert.py npc1.tab dialog.txt
#
# Blakserv admin command:
# read dialog.txt (file needs to be in server run directory)

import sys

def usage():
    print('Usage:')
    print(sys.argv[0] + ' input_file output_file')
    sys.exit()

def main():
    if len(sys.argv) < 3:
        usage()

    input_filename = sys.argv[1]
    output_filename = sys.argv[2]

    output_file = open(output_filename, "w")
    
    with open(input_filename, "r") as input_file:
        for line in input_file:
            fields = line.split('\t')

            if fields[2]:
                command1 = f"send o 0 speechlibtrigger who class {fields[0]}"
                command2 = f"send o 0 speechlibquote who class {fields[0]} quote q {fields[3]}"
                if fields[1]:
                    command1 += f" mood int LIBRES_{fields[1]}"
                if fields[2]:
                    command1 += f" trigger q {fields[2]}"
                output_file.write(command1)
                output_file.write("\n")
                output_file.write(command2)
                output_file.write("\n")
            else:
                command1 = f"send o 0 randomlibquote who class {fields[0]}"
                if fields[1]:
                    command1 += f" mood int LIBRES_{fields[1]}"
                command1 += f" quote q {fields[3]}"
                output_file.write(command1)
                output_file.write("\n")
            
    output_file.close()

main()
